packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -433,133 +669,242 @@ fn array_to_user_warnings(array: mv8::Value<'_>) -> Vec<UserWarning> {
-            let errors = array_to_user_errors(errors);
+fn get_js_error<'s>(
+    scope: &mut v8::HandleScope<'s>,
+    return_val: v8::Local<'s, v8::Object>,
+) -> Option<JSRunnerError> {
+    tracing::debug!("Collecting JS errors");
+    let user_errors = if let Some(user_errors) = v8::String::new(scope, "user_errors") {
+        user_errors
+    } else {
+        return Some(JSRunnerError::V8("Could not create String".to_string()));
+    };
+
+    if let Some(errors) = return_val.get(scope, user_errors.into()) {
+        if !errors.is_null_or_undefined() {
+            let errors = array_to_user_errors(scope, errors);
             if !errors.is_empty() {
-                return Some(Error::User(errors));
+                return Some(JSRunnerError::User(errors));
             }
         }
     }
 
-    if let Ok(mv8::Value::String(e)) = return_val.get("pkg_error") {
-        // TODO: Don't silently ignore non-string, non-null-or-undefined errors
-        //       (try to convert error value to JSON string and return as error?).
-        return Some(Error::Package(PackageError(e.to_string())));
+    let pkg_error = if let Some(pkg_error) = v8::String::new(scope, "pkg_error") {
+        pkg_error
+    } else {
+        return Some(JSRunnerError::V8("Could not create String".to_string()));
+    };
+
+    let runner_error = if let Some(runner_error) = v8::String::new(scope, "runner_error") {
+        runner_error
+    } else {
+        return Some(JSRunnerError::V8("Could not create String".to_string()));
+    };
+
+    if let Some(e) = return_val.get(scope, pkg_error.into()) {
+        // Even though rusty_v8 returns an Option, if the object does not have the property the
+        // result will be Some(undefined)
+        if !e.is_undefined() {
+            let e: v8::Local<'s, v8::String> = match e.to_string(scope) {
+                Some(e) => e,
+                None => {
+                    return Some(JSRunnerError::V8(
+                        "Could not convert e to String".to_string(),
+                    ));
+                }
+            };
+
+            // TODO: Don't silently ignore non-string, non-null-or-undefined errors
+            //       (try to convert error value to JSON string and return as error?).
+            return Some(JSRunnerError::Package(PackageError(
+                e.to_rust_string_lossy(scope),
+            )));
+        }
     }
 
-    if let Ok(mv8::Value::String(e)) = return_val.get("runner_error") {
-        // TODO: Don't ignore non-string, non-null-or-undefined errors
-        return Some(Error::Embedded(e.to_string()));
+    if let Some(e) = return_val.get(scope, runner_error.into()) {
+        // Even though rusty_v8 returns an Option, if the object does not have the property the
+        // result will be Some(undefined)
+        if !e.is_undefined() {
+            let e: v8::Local<'s, v8::String> = match e.to_string(scope) {
```

Comment:
Maybe we want to use `if let` here?

Reply:
https://github.com/hashintel/hash/pull/414/commits/45723b0145a7715a94ee28f82b5da9b70a3e5f84